### PR TITLE
Cupcake parallel

### PR DESCRIPTION
**Pull request by bmcase**
Cupcake parallel struct as example with tests. Also PS3I outline

### CODE REVIEW

File: examples/parallel.rs [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+
+impl ByteBuffer {
+    pub fn from_slice(v: &[u8]) -> ByteBuffer {
+        ByteBuffer { buffer: v.to_vec() }
+    }
+}
+
+impl Display for ByteBuffer {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), Error> {
+        //todo : fix me with proper formatting
```
Comment (haochenuw) [inline examples/parallel.rs:29]: is this still a todo? if not, comment can be removed

File: examples/parallel.rs [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+
+
+// a struct for Parallel Cupcake operations
+
+    // usage for main method PS3I outline:
+    //      cargo run --example parallel
+    //      cargo run --release --example parallel
+    // usage for CupcakeParallel tests
+    //      cargo test --examples parallel
+    //      cargo test --release --examples parallel
```
Comment (haochenuw) [inline examples/parallel.rs:55]: We could move this into the readme file

File: examples/parallel.rs [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+use cupcake::FVCiphertext;
+use cupcake::FVPlaintext;
+use cupcake::integer_arith::scalar::Scalar;
+use cupcake::SecretKey;
+
+use rayon::iter::IntoParallelIterator;
+use rayon::iter::ParallelIterator;
+use rayon::iter::IndexedParallelIterator;
+
+use serde::{Deserialize, Serialize};
```
Comment (haochenuw) [inline examples/parallel.rs]: We have added support for custom serialization. Could we use that?

File: examples/parallel.rs [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+        let scheme_ = &self.scheme;
+        let sk_ = &self.sk;
+        payload.into_par_iter().map(|item| {
+                let pt = scheme_.decrypt(&scheme_.from_bytes(&(item.buffer)), &sk_);
+                return pt;
+            })
+            .collect::<Vec<Vec<u8>>>()
+    }
+
+    // also deserializes and serializes at both ends.
```
Comment (haochenuw) [inline examples/parallel.rs]: comment a bit unclear to me

File: examples/parallel.rs [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+                let mut ct:FVCiphertext<Scalar> = scheme_.from_bytes(&ct_bytes.buffer);
+                &self.scheme.sub_plain_inplace( &mut ct, pt); // subtracts pt into ct
+                ByteBuffer {
+                    buffer: ct.to_bytes()
+                }
+            })
+            .collect::<Vec<ByteBuffer>>()
+    }
+
+    // would it be more efficient to pass by mutable reference and have no return??
```
Comment (haochenuw) [inline examples/parallel.rs]: To answer the question, it depends on whether the ciphertext is still used for other things.

File: examples/parallel.rs [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+    pub fn simple_plaintext_ciphertext_add(&self, ct:&mut FVCiphertext<Scalar>, pt: &FVPlaintext<u8>  ){
+        let scheme_ = &self.scheme;
+        scheme_.add_plain_inplace( ct, pt);
+    }
+}
+
+
+
+fn main() {
+    // Oultine of PS3I HE operations using cupcake
```
Comment (haochenuw) [inline examples/parallel.rs]: typo -> Outline

File: src/traits.rs [hunk 7]
```diff
@@ -40,6 +40,8 @@ pub trait PKEncryption<CT, PT, SK>: SKEncryption<CT, PT, SK> {
 pub trait CipherPlainAddition<CT, PT>: {
```
Comment (haochenuw) [inline src/traits.rs]: let's update the name to cipher plain AddAndSubtract

File: src/lib.rs [hunk 8]
```diff
@@ -256,6 +256,14 @@ where
             *ct_coeff = T::add_mod(ct_coeff, &temp, &self.q);
         }
     }
+
+    // not implemented yet
```
Comment (haochenuw) [inline src/lib.rs]: comment not needed any more? 
Also let's add tests to sub_plain_inplace

File: Cargo.toml [hunk 9]
```diff
@@ -12,7 +12,10 @@ description = "An implementation of a lattice-based additive homomorphic encrypt
 [dependencies]
 rand = "0.5.5"
 modinverse = "0.1.1"
-
+rayon = "1.3.0"
+#prelude = "0.2.1"
+serde = {version = "1.0.104", features = ["derive"] }
```
Comment (shubho) [inline Cargo.toml]: Do we need `byte_buffer`?

File: examples/parallel.rs [hunk 10]
```diff
@@ -0,0 +1,356 @@
...
+use cupcake::integer_arith::scalar::Scalar;
+use cupcake::SecretKey;
+
+use rayon::iter::IntoParallelIterator;
+use rayon::iter::ParallelIterator;
+use rayon::iter::IndexedParallelIterator;
+
+use serde::{Deserialize, Serialize};
+use std::fmt::{Debug, Display, Error, Formatter};
+
```
Comment (shubho) [inline examples/parallel.rs:16]: `ByteBuffer` is already serialized as a `vec<u8>` so it shouldn't need a `Serialize` or `Deserialize` trait

File: Cargo.toml [hunk 11]
```diff
@@ -12,7 +12,10 @@ description = "An implementation of a lattice-based additive homomorphic encrypt
 [dependencies]
 rand = "0.5.5"
 modinverse = "0.1.1"
-
+rayon = "1.3.0"
```
Comment (shubho) [inline Cargo.toml:15]: Remove comment?

File: examples/parallel.rs [hunk 12]
```diff
@@ -0,0 +1,356 @@
...
+use cupcake::DefaultShemeType;
+use cupcake::FVCiphertext;
+use cupcake::FVPlaintext;
+use cupcake::integer_arith::scalar::Scalar;
+use cupcake::SecretKey;
+
+use rayon::iter::IntoParallelIterator;
+use rayon::iter::ParallelIterator;
+use rayon::iter::IndexedParallelIterator;
+
```
Comment (shubho) [inline examples/parallel.rs:14]: I don't think we need `serde` since we don't need those traits

File: examples/parallel.rs [hunk 13]
```diff
@@ -0,0 +1,356 @@
...
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+use cupcake::traits::*;
+use cupcake::DefaultShemeType;
+use cupcake::FVCiphertext;
+use cupcake::FVPlaintext;
+use cupcake::integer_arith::scalar::Scalar;
+use cupcake::SecretKey;
+
```
Comment (shubho) [inline examples/parallel.rs:12]: Can be formatted as `use rayon::iter::{IntoParallelIterator, ParallelIterator, IndexedParallelIterator};`

File: Cargo.toml (hunk 9 above)
Comment (bmcase) [reply Cargo.toml]: nope, removed

File: examples/parallel.rs (hunk 1 above)
Comment (bmcase) [reply examples/parallel.rs:29]: @shubho , I think I copied this todo you left. should we change?

File: examples/parallel.rs (hunk 4 above)
Comment (bmcase) [reply examples/parallel.rs]: changed to :     // method takes serialized input and serializes the output

File: README.md [hunk 14]
```diff
@@ -25,15 +25,15 @@ cargo build --release
 ## Examples
 
 Several examples are included in `examples/<name>.rs`, and can be run via
-`cargo run --example <name>`
+`cargo run --example <name>` or `cargo run --release --example <name>`
```
Comment (haochenuw) [inline README.md:28]: might be good to say a few words about what --release does

File: examples/parallel.rs [hunk 15]
```diff
@@ -0,0 +1,351 @@
...
+// a struct for Parallel Cupcake operations
+
+    // usage for main method PS3I outline:
+    //      cargo run --example parallel
+    //      cargo run --release --example parallel
+    // usage for CupcakeParallel tests
+    //      cargo test --examples parallel
+    //      cargo test --release --examples parallel
+
+pub struct CupcakeParallel {
```
Comment (haochenuw) [inline examples/parallel.rs:57]: Can we put the the struct and implementations in lib.rs? The goal is to keep example files strictly for examples.

Comment (haochenuw) [reply examples/parallel.rs:57]: Also let's have some high level comments documenting the struct and its API

File: examples/parallel.rs [hunk 16]
```diff
@@ -0,0 +1,351 @@
...
+        let scheme_ = cupcake::default();
+        let (pk_, sk_) = scheme_.generate_keypair();
+        CupcakeParallel {
+            scheme: scheme_,
+            pk: pk_,
+            sk: sk_
+        }
+    }
+
+    pub fn parallel_encrypt_serialize(&self, plaintext:&Vec< Vec<u8> >) -> Vec<ByteBuffer> {
```
Comment (haochenuw) [inline examples/parallel.rs:74]: nit: encrypt_and_serialize?

File: examples/parallel.rs [hunk 17]
```diff
@@ -0,0 +1,351 @@
...
+                &self.scheme.sub_plain_inplace( &mut ct, pt); // ct := ct - pt
+                &self.scheme.rerandomize(&mut ct, &pk_);
+                ByteBuffer {
+                    buffer: ct.to_bytes()
+                }
+            })
+            .collect::<Vec<ByteBuffer>>()
+    }
+
+        // also deserializes and serializes at both ends.
```
Comment (haochenuw) [inline examples/parallel.rs:114]: nit: alignment

File: examples/parallel.rs [hunk 18]
```diff
@@ -0,0 +1,351 @@
...
+                let mut ct:FVCiphertext<Scalar> = scheme_.from_bytes(&ct_bytes.buffer);
+                &self.scheme.add_plain_inplace( &mut ct, pt); // adds pt into ct
+                ByteBuffer {
+                    buffer: ct.to_bytes()
+                }
+            })
+            .collect::<Vec<ByteBuffer>>()
+    }
+
+    pub fn simple_encrypt(&self, plaintext: &FVPlaintext<u8>) -> FVCiphertext<Scalar> {
```
Comment (haochenuw) [inline examples/parallel.rs:168]: how is this different from scheme.encrypt?

File: examples/parallel.rs [hunk 19]
```diff
@@ -0,0 +1,351 @@
...
+    pub fn simple_plaintext_ciphertext_add(&self, ct:&mut FVCiphertext<Scalar>, pt: &FVPlaintext<u8>  ){
+        let scheme_ = &self.scheme;
+        scheme_.add_plain_inplace( ct, pt);
+    }
+}
+
+
+
+fn main() {
+    // Outline of PS3I HE operations using cupcake
```
Comment (haochenuw) [inline examples/parallel.rs:184]: A high level question: what is the example supposed to demonstrate? If it is for parallel operations, would it be better to benchmark them?

File: examples/parallel.rs [hunk 20]
```diff
@@ -0,0 +1,351 @@
...
+    let v_p_double_prime = parfv_partner.parallel_deserialize_decrypt(&ctv_p_double_prime);
+
+    println!("decrypted v_p_double_prime = v_p - r_p: ");
+    smartprint2(&v_p_double_prime);
+
+
+}
+
+
+#[cfg(test)]
```
Comment (haochenuw) [inline examples/parallel.rs:251]: similar to above, we can put it in the library create instead of example

**Pull request closed without merging**